KerbalStuff/thumbnail.py

```diff
@@ -1,12 +1,62 @@
 import os.path
-from typing import Tuple
 
 from PIL import Image
 
+from KerbalStuff.config import _cfg, _cfgi, site_logger
 
-def create(imagePath: str, thumbnailPath: str, thumbnailSize: Tuple[int, int]) -> None:
-    if not os.path.isfile(imagePath):
+
+def create(background_path: str, thumbnail_path: str) -> None:
+    if not os.path.isfile(background_path):
         return
-    im = Image.open(imagePath)
-    im.thumbnail(thumbnailSize, Image.ANTIALIAS)
-    im.save(thumbnailPath, 'jpeg', quality=50, optimize=True)
+
+    size_str = _cfg('thumbnail_size')
+    if not size_str:
+        size_str = "320x195"
+    size_str_tuple = size_str.split('x')
+    size = (int(size_str_tuple[0]), int(size_str_tuple[1]))
+
+    quality = _cfgi('thumbnail_quality')
+    if not quality or 0 > quality > 100:
+        quality = 80
+
+    im = Image.open(background_path)
+
+    x_ratio = abs(im.width/size[0])
+    y_ratio = abs(im.height/size[1])
+
+    if x_ratio < y_ratio:
+        im = im.resize((size[0], round(im.height/x_ratio)), Image.LANCZOS)
+    else:
+        im = im.resize((round(im.width/y_ratio), size[1]), Image.LANCZOS)
+
+    box_left = 0.5 * (im.width - size[0])
+    box_upper = 0.5 * (im.height - size[1])
+    box_right = 0.5 * (im.width + size[0])
+    box_lower = 0.5 * (im.height + size[1])
```

Comment:
Can you please add a comment explaining what these calculations mean at a high level? `box_right` and `box_lower` are the averages of the actual and configured sizes, but I'm not having much luck figuring out `box_left` and `box_upper`. Are we trying to get a centered box inside the newly resized image?

Reply:
Yep, this tries to crop out an image that has the configured size, from the centre of the base image.
`image.crop()` basically takes the upper left and lower right corener, but not as two tuples but instead 4 separate values: https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.crop

I also think I have to make sure that they're actually integers.